### ISSUE TITLE
Add  context to DiagnosticSource

### ISSUE THREAD

**Issue by lmolkova** [dotnet/corefx#15023]
# Motivation
Our goal is to enable telemetry event correlation.  To do this we need incoming requests to be tagged with an ID (and potentially other information), and have this information 'flow' through (be available) to all code run in response to that request and in particular allow this information to be propagated to any outgoing requests.    

This class is meant to effectively augment existing DiagnosticSource telemetry, and be compatible/leveragable from using the existing ILogger and EventSource based logging as well.    The result is that all these logging mechanisms will now have 'end-to-end' event correlation (every event has an activity, and for any event you can find the chain of activities that 'caused' it).  

# Concepts 
The main new code is the Activity class.   Conceptually an Activity is something that starts and stops (thus it has a start time and duration), and potentially has a 'Parent' that 'caused' it.   In addition it has  two sets of arbitrary key-value pairs.    The first is called 'Tags', which are intended to be logged with the activity but do NOT propagate to children.  The second is 'Baggage' that DOES propagate to children.   
Tags are intended to augment telemetry, baggage is intended to control telemetry.  

# Usage scenario: 
Modify existing Incoming HTTP Request telemetry code (e.g. in ASPNet.Hosting) to 
 -  Extract data (e.g. an ID) from the HTTP header
  - Create a new a new Activity which has this ID as a parent and stores any other telemetry data values from HTTP header.   
  - Make the activity current with the Activity.Current and log the telemetry with the DiagnosticSource.Start() API.  
 - Now at any point, the telemetry context is available in Activity.Current.

Modify the existing Outgoing HTTP request telemetry (e.g. in System.Net.Http) to 
  - Looks at Activity.Current and injects the current ActivityID and Baggage into the outgoing HTTP header.  

This feature is focused on context itself and does not cover incoming/outgoing operations handling (they are in different DLLs and will be covered in subsequent pull requests).  
The API is inspired by [opentracing.io](http://opentracing.io/) standard ([API specification](https://github.com/opentracing/specification/blob/master/specification.md)) and implements following concepts:

# Proposed API
```c#
namespace System.Diagnostics
{
    public class Activity
    {
        //name of operation
        public string OperationName { get; }

       //id of the activity. Propagated to dependencies, so parent-child relationship of activities could be logged
        public string Id { get; private set; }

        // activity start timestamp
        public DateTime StartTimeUtc { get; private set; }

        //duration for stopped activity
        public TimeSpan Duration { get; private set; }

        //reference to parent activity if any;
        public Activity Parent { get; private set; }

        //represents Id of parent activity, may be passed from external service
        public string ParentId { get; private set; }

        // list of key-value pairs which may be used for logging. E.g. HTTP Method and Path. Users can enumerate tags and add new tags
        public IEnumerable<KeyValuePair<string, string>> Tags {get;}
       
       //list of key value pairs to be propagated to downstream services. E.g. correlation-id. Users can enumerate baggage, request baggage by name and add new items
        public IEnumerable<KeyValuePair<string, string>> Baggage {get;}

       //Current Activity for the current thread.  This flows across async calls.
        public static Activity Current;

        public Activity(string operationName);
        public Activity AddTag(string key, object value);
        public Activity AddBaggage(string key, string value);
        public Activity SetParentId(string parentId);
        public Activity SetStartTime(DateTime startTimeUtc);
        public Activity SetEndTime(DateTime endTimeUtc);
        public string GetBaggageItem(string key);        

        //Sets activity parent and Activity.Current
        public Activity Start();

        //Restores Activity.Current to parent activity
        public void Stop();
    }
    
   // wrapper methods in DiagnosticSource for activity Start and Stop which writes corresponding events to DiagnosticSource
    public abstract partial class DiagnosticSource // in addition to existing methods
    {
        public Activity StartActivity(Activity activity, object args);
        public void StopActivity(Activity activity, object args);        
    }
}
```

# Usage Sample

```c#
namespace App
{
    public void OnIncomingRequest(DiagnosticListener httpListener, HttpRequest request)
    {
        // This code represents code in say web service's hosting layer that receives an 
        // Incoming HTTP from outside the process.  We expect to instrument this code 
        // to accomplish the following three things
        //    1) Create a new activity that will track through the system as the request is processed
        //    2) Look for items in the HTTP request that can be used to correlate this incoming
        //        request with other processing on other machines.  These should be put into
        //        the newly created Activity so it is available for all subsequent logging.  
        //    3) Log the start and stop of this activity that was created.  
        // 
        // In this example we look for the 'x-ms-request-id' as a HTTP header that identifies 
        // a ID unique ID and passes along any values with the 'x-ms-baggage' prefix.    
        // 
        if (httpListener.IsEnabled("Http_In"))
        {
            var activity = new Activity("Http_In");
            
            //parse request headers and add tags and baggage
            activity.SetParentId(request.headers["x-ms-request-id"])
            foreach (var header in context.Request.Headers)
                    if (header.Key.StartsWith("x-ms-baggage-")
                        activity.AddBaggage(header.Key, header.Value);

            //start activity
            httpListener.StartActivity(activity, context);
            try {
                 //process request ...
            } finally {
                //stop activity
                httpListener.StopActivity(activity, context); 
            }       
        }
    }
    
    public void OnOutgoingRequest(DiagnosticListener httpListener, HttpRequestMessage request)
    {
        // This code represents code in the framework where an outgoing HttpRequest is being 
        // sent to another machine.   The goal of this code is to create and log an activity associated
        // with doing this as well as to add the ID and any Activity Baggage to the outgoing HTTP
        // header.  
        if (listener.IsEnabled("Http_Out"))
        {
            var activity = new Activity("Http_Out");

            //start new activity, Parent for this activity is assigned to Activity.Current
            listener.StartActivity(activity,  new { Request=request) });
            try {
                // Attach our ID and Baggage to the outgoing Http Request
                request.Headers.Add("x-ms-request-id", activity.Id);
                foreach (var baggage in activity.Baggage)
                    request.Headers.Add(baggage.Key, baggage.Value);
         
               // Actually perform the Http Write Request, get the response 
           }  finally { 
                listener.StopActivity(activity, {new { Response=reponse });
          }
        }
    }
    
    public void Startup()
    {
       // This is code that shows usage to get at the logging (and activity) information
       // logged by the code above.   This follows normal DiagnosticListener conventions.  
       // In which you subscribe to and new DiagnosticListener.
       // (We assume the DiagnosticListener we used above is called 'System.Net.Http')
       // From there we can subscribe to its events.  
       // Events will be fired on every Start() and Stop() call above, and the Activity.Current 
       // will be accurate during those callbacks.  Thus you have all the information passed int
       // the Start() or Stop() call as well as all information from Activity.Current available
       // in the callback. 
       DiagnosticListener.AllListeners.Subscribe(delegate (DiagnosticListener listener)
       {
            if (listener.Name == "System.Net.Http")
            {
                listener.Subscribe(delegate (KeyValuePair<string, object> value)
                {
                     // In this demo we simply log to the console what things are happening. 
                     Console.WriteLine("Event: {0}  Current Operation: {1} Current ID {2}", 
                           value.Key, {Activity.Current.OperationName}, {Activity.Current.Id});
                });
            }
        }
    }
}
```

# Notes
We want to bring Activity to ASP.NET 4.5 and ASP.NET Core users which require DiagnosticSource for NET 4.5.
Netstandard1.1 is not good enough because it does not implement nethier AsyncLocal nor CallContext,

**Comment by cwe1ss** [comment]
I implemented the C# version of opentracing.io so I have great interest in this topic. 

Any chance we can talk about making this implement the opentracing API in a more compliant way? It would be bad if we had a "almost"-implementation in the framework and a separate opentracing library.

I'd be happy to connect you with the people behind opentracing to chat about possible questions and concerns.

**Comment by karelz** [comment]
cc @vancem

**Comment by vancem** [comment]
@cwe1ss  

We certainly want to extract value out of the opentracing.io standard if we can, and are certainly open to suggestions.   We in fact started with opentracing.io as a model.    The 'Activity' class is very much like the OpenTracing 'Span' class (calling it that however, is a bit problematic for us, as that name has been 'claimed' for class that will represent a slice of an array (yes they would be in different namespaces, but the confusion would still persist).   

So the answer to your suggestion about getting questions/concerns addressed, is, yes, we would like to address any concerns.  Certainly we want opentracing.io 'back ends' to be able to consume this information without any 'pain'.   

A major goal we had however was to NOT generate yet another way of logging information.    The idea is that the existing mechanisms for logging in .NET (DiagnosticSource, ILogger, EventSource), would simply share a correlation scheme (Activity).    Our expectations is that the framework would be instrumented with DiagnosticSource and user code with any of the others (e.g. ILogger), and that they can all work together (including sending their information it opentracing.io compliant loggers complete with correlation information.   

This makes the design different than a classic opentracing.io implementation because openTracing wants to take over as the logger (it is both the logging system and the correlation system).  

One way of looking at the design that is being proposed is that the combination of DiagnosticSource and Activity basically produces an output stream that should flow naturally into a opentracing.io back end (certainly the models are compatible).    

But it should also be true that logging done with ILogger would also produce a stream that would also naturally flow to an opentracing.io back end (as well as ILogger-native back ends etc).    The design allows for this.  

I think it is a fair point that we should confirm that and end-to-end example using openTracing back end should  work well before the design is solidified.   We think this is true, but we have not done that validation yet.   

The reason we believe we have a good shot at 'just working' with opentracing.io is because we are trying to be flexible and minimal, and try to make as few assumptions as possible to make forward progress.  
Right now the forward progress we want is to have enough correlation such that activity IDs can be injected and extracted from HTTP headers (so we can hop across machines).      One of the few assumptions we make is that things in the framework (in particular the System.Net.Http stack) are instrumented with DiagnosticsSource (in the framework, we don't take dependencies on other loggers).  
The framework has to pick a logging system that it relies on, and DiagnosticSource was created FOR THAT PURPOSE, so staring with that assumption seems very reasonable (and any other logging system can be hooked up on the other side of DiagnosticSource).  

Thus the short term goal of getting correlation reduces to adding 'just enough' extra infrastructure into DiagnosticSource (the Start and Stop calls), that hook into a GENERAL PURPOSE, 'causality flow' class (Activity, that happens to be stand-alone and independent of any particular logging system), seems like a step that 'can't be wrong'.   This is the step we are making above.     The intent is that all other 'glue' can be put behind DiagnosticSource (which really does not do much and thus is very flexible).  

Indeed you will see that Activity' has a model that is basically identical to opentracing.io's Span (it has a start time and duration, operation name, an ID and a Parent ID).  In fact we took the concept of Tags and Baggage from Span.    Thus so far, we should be in good shape to interoperate with opentracing.io.   

However there are plenty of 'devils' in the details, and feedback and discussion are more than welcome.

**Comment by cwe1ss** [comment]
Hi Vance, I'm happy to hear that there's interest in making it as compatible as possible.

I fully agree that there shouldn't be yet another logging solution in the framework! We've already had a few discussions about the existing logging systems in .Net in other github issues and I'm still very thankful for the valuable insight you've provided there!

Traditional logging has been added to Opentracing just recently and the spec for it is still very new. IMO it's also missing some important things which are already solved with ILogger and EventSource (e.g. explicit handling of exceptions). Therefore I also agree, that it's easy to consume these existing loggers in an OpenTracing bridge and it doesn't need to be on `Activity`.

Given that DiagnosticSource already exists, it seems fair and good to use it as the foundation for this feature. actually, in my WIP of an OpenTracing instrumentation for the .NET framework, I'm using it as well. You can find my code here: https://github.com/cwe1ss/opentracing-contrib-dotnet

I have the following feedback/concerns/questions regarding the proposed API:

#### Naming
Independently of opentracing, I think using the name `Span` for the new array slice in the root `System` namespace is not good. Span is a very generic word and the concept already exists in many areas (e.g. HTML, Xamarin Forms). I've raised my concerns regarding this in https://github.com/dotnet/coreclr/issues/7636. 

Even if the new array `Span` was renamed, I'm not sure if this `Activity` should be renamed to `Span` then. This would depend on how compatible the API would be. But since there won't be 100% coverage, users can't use the official opentracing docs anyway.

#### Tracer interface - Inject and Extract
OpenTracing has the very important concept of injecting and extracting span/trace metadata (ID, baggage, ...) into carrier objects (e.g. Http request). The goal of Opentracing is to be vendor neutral, this also means instrumentation code can't know the name of the HTTP headers and tracers might also have to add additional headers which are not part of the baggage. 

Given your example code, would every instrumentation code in the world have to hard code the string `x-ms-request-id`, etc?!

Also, for example, a tracer like Application Insights must propagate whether the current trace is sampled or not (to ensure every sub request of a trace is captured, even if Application Insights is configured to only capture 1% of all requests)

In other words, providing a hook/indirection for how data is propagated onto carrier objects is critical IMO. Otherwise there's no way to add additional functionality without touching each instrumentation (which is impossible once it's used outside of the framework as well).

#### Span / Activity

The API here seems similar, but as you've said the devil is in the details. Some first concerns:

* Tags only accept string. However, it's important to be able to pass bool, int and double values as well. Not having support for this will lead to conversion issues. Also it might be wasteful to convert everything to string.

* Start / Stop / Dispose: currently, not all DiagnosticSource usages reliable call an event in case of failure (e.g. HttpClient https://github.com/dotnet/corefx/issues/13172). It would be great if DiagnosticSource could use IDisposable and using statements for these kind of calls. If not, reverting Activity.Current is problematic and you can't mark activities as failed.

There are probably more things - I'd go into more details once we are further in the discussion.

I'd happy to work with you on this topic! However I'm traveling through Colombia until mid of February 😃 , so I can't do any programming before that.

**Comment by vancem** [comment]
Thanks for the feedback Christian.      

Naming:   I personally have no strong preference here.  It will depend on what happens with the array slicing, and I will bring it up with people who deal with naming more.

Inject/Extract:  I do see there is abstraction here, and we did think about this in the design, but I was having a hard time coming up with concrete value of the abstraction (exactly who is being hidden from whom and why is that valuable?)    I am open to persuasion, but if I am going morph/copy things I want to understand why I am doing it.   It seems like EVERY transport technology has the same problem that HTTP has (if you allow arbitrary keys-value pairs, you need a prefix).  Well given that HTTP is likely to be the 95+% case, the SIMPLEST / MOST EFFICIENT solution is have Activity REQUIRE a prefix that works well for HTTP and is likely to work well for ANY transport, and then have no morphing into and out of the header (you can see the example code above, and it is pretty tight).    

Anyway, this issue SEEMS to be roughly independent of Activity (Inject and Extract can be implemented as extension methods that use the current public API of Activity).    This code will live elsewhere (after all we don't want DiagnosticSource to have dependencies on HTTP) so strictly speaking this issue does not block this part of the design.  

Tags:  changing it from string->string to string->object is worth considering.   

IDispose for Start:   Definitely worth considering.   We have not done it so far because when your 'stop' activity needs to pass more information it does not fit the dispose pattern.   That made us pause.

**Comment by karelz** [comment]
@lmolkova @vancem please follow [API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md)

**Comment by karelz** [comment]
@vancem here's background on [triage and issues](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#triage-rules---simplified), I'll be happy to give you crash course as we didn't triage Tracing area yet (cc @valenis)

**Comment by vancem** [comment]
@karelz I have marked the issue as 'API-read-for-review'.    @lmolkova  and I have already gone through several rounds of updates, and while there may be a tweek or two left, its rough shape and usage should be stable now.   

@lmolkova  has also submitted a correspoinding pull request #15103  which was done not to circumvent the review process, but to provide additional detail.   In particular the user's guide 

https://github.com/lmolkova/corefx/blob/10512ae7ca7bb6a8efd7b1b32d16bd7fe25a30ad/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md

will be helpful for the review as well as the Activity.cs source file because it has detailed 
descriptions of each of the new APIs.  

https://github.com/lmolkova/corefx/blob/10512ae7ca7bb6a8efd7b1b32d16bd7fe25a30ad/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

**Comment by karelz** [comment]
Perfect, we should cover it in one of the next API review meetings (we have 15 issues now, so it may take 2 weeks for us to get to it).
The API proposal is typically sufficient for API review.

BTW: I skimmed through it - the name 'Baggage' seems a bit weird to me. Did you consider any alternatives?

**Comment by danmosemsft** [comment]
Baggage seems to be the opentracing.io term: https://github.com/opentracing/opentracing.io/blob/cccdae2f2a09969945a857ae73328b4b4ed1d11f/spec.md

**Comment by karelz** [comment]
Thanks for explanation. It makes sense to be consistent with open standards ...

**Comment by terrajobst** [comment]
We reviewed this today with @vancem, here is some high-level feedback:

 * Shouldn't use fluent interface design as it's not clear whether the object is immutable or mutable
 * `Tags` and `Baggage` should probably be a collection type that offers an add so that collection initializer syntax can be used and typical expectations are met, i.e. `activity.Tags.("x", "y")` works. One option would be to use tuples, e.g. `public ICollection<(string, object)> Tags { get; }`
* `Tags` and `Baggage` should use the same type for the value. Ideally they would use `string` to make sure the type can trivially serializable
* The extension type seems to add little value;  we should merge the methods with `Activity`
* We should consider making `Start` and `Stop` methods non-static
* Should `Start` and `Stop` be called `LogStart` and `LogStop`?

**Comment by vancem** [comment]
@terrajobst, @KrzysztofCwalina  
In reply to review feedback
 > Shouldn't use fluent interface design as it's not clear whether the object is immutable or mutable

I agree that it should be clear that the object is mutable.   I propose a simple renaming

   WithTag -> AddTag
   WithBaggage -> AddBaggage
   All other With* -> Set*
I note that in https://en.wikipedia.org/wiki/Fluent_interface there first example was a mutable object.

> Tags and Baggage should probably be a collection type that offers an add so that collection initializer syntax can be used and typical expectations are met, i.e. activity.Tags.("x", "y") works. One option would be to use tuples, e.g. public ICollection<(string, object)> Tags { get; }

I would like to keep this unchanged.  The argument is that a simpler, more efficient (fewer objects) implementation is possible without the extra conceptual object (the ICollection).    The Tags and Baggage ARE likely to be on the hot path but the Adding of tags and baggage are likely to be rare, so optimizing a little for efficiency over convenience is reasonable.  

> Tags and Baggage should use the same type for the value. Ideally they would use string to make sure the type can trivially serializable

We will change Tags to be have a 'string' value.   

> The extension type seems to add little value; we should merge the methods with Activity

We can make the Start and Stop() be direct methods on DiagnosticSource.   We can't easily merge these into Activity, however because Activity does not Know about DiagnosticSource (but it is reasonable for DiagnosticSource to know about Activity, which is what will happen).

> We should consider making Start and Stop methods non-static 

We can do that.   The original idea was that only static methods touch the (static) 'Current' property.   It also makes clearer that these methods are not like the other methods (they touch global context)  However we can describe it in documentation.  I should note however, that these methods are NOT likely to used frequently (we expect them only to be used in logger 'Start' and 'Stop' routines like the one for DiagnosticSource.  

Should Start and Stop be called LogStart and LogStop?

I assume you mean the Start() and Stop() on DiagnosticSource (since that is the one that Logs).  The API that logs in DiagnosticSource is called 'Write' so using 'Log' in some APIs but 'Write' in others seems odd.    I would not have a problem with 'WriteStart' and WriteStop'   But I also don't mind 'StartActivity' and 'StopActivity' (but I am also OK with leaving them as is.

Note that we don't expect these APIs to be used commonly, so we should not sweat too much over API design assuming it is 'close enough'.

**Comment by terrajobst** [comment]
@vancem 

Your proposals sound reasonable. Could you update the proposal on the top to reflect these adjustment? I think we can close on this without a meeting.

@dotnet/fxdc: any concerns?

**Comment by lmolkova** [comment]
@terrajobst 

Proposal is updated, as well as  PR #15103 in case you want to take a look.

**Comment by terrajobst** [comment]
@karelz , please update to api-approved unless someone else from @dotnet/fxdc pushes back.

**Comment by karelz** [comment]
Done. If there's pushback in next couple of days, we can revert it.

**Comment by davkean** [comment]
Why DateTime over DateTimeOffset? Then no explicit "ask" for Utc is required.

**Comment by cwe1ss** [comment]
I still have many concerns with the current proposal. 

Please keep in mind that .NET is a baby in the distributed tracing world and interoperability should therefore be a main concern. What if a Go client wants to call a .NET web API? By hard coding keys like "x-ms-*" into core parts of the .NETframework you make this a proprietary solution that has little to no value for any mixed-tech company.

There has been a lot of discussion in the opentracing repos (opentracing-java and opentracing-go might be the most useful) regarding many of the things here (general points and API details) and it would be great if we could consider/discuss them here as well.

I'm really worried that this ends up pretty much useless in a real-world scenario and that I have to keep most of the code in opentracing-csharp - which means that everything will be instrumented/timed multiple times etc. and therefore waste memory and CPU cycles.

How fast do you need this implemented? I'd love to work on this with you (and hopefully other opentracing contributors), but since I'm on a longer holiday, I can't do much before Feb 20th.

Could you please keep the API discussion open until then?

**Comment by cwe1ss** [comment]
> Why DateTime over DateTimeOffset? Then no explicit "ask" for Utc is required.

IMO DateTime is better. There's no point in tracking time zones. Timing should always be in UTC. Ideally (IMO), the code would throw if `Kind != Utc`.

Fyi, I looked into this topic in (a lot) more detail in https://github.com/opentracing/opentracing-csharp/issues/4

**Comment by karelz** [comment]
@cwe1ss thanks for offer of help - I let @vancem @lmolkova comment if they can wait until 2/20 for deeper discussion.
The good news is that before we ship publicly (in 2.0), we can change our mind on API shape, implementation, etc. If we don't believe we have final API shape, we can always back out the API surface before shipping. The only question is how it affects @lmolkova's deliverables and schedule (I don't have visibility there).

**Comment by vancem** [comment]
I agree with @cwe1ss.   I believe it is standard in the server world to use UTC uniformly.   This is because it is SIMPLE (the notion of a point in time, independent of where you are) and EFFICIENT (smaller, less logic).   DateTimeOffset makes much more sense when you are interacting with humans (who want their time in the local zone.   Thus it would make sense for the DISPLAY tools, but not really the collecting infrastructure (which is what this is.  

On the issue of names in HTTP headers.   We are aware of this problem and in fact already have a notion of mapping that we are thinking about.    We are open to other ideas as well.    The code above is JUST about Activity (not the code in HTTP stacks), and the example code is a simplification, so don't get hung up about that at this point.   

We need to proceed with implementation with due speed (Frankly we wanted this checked in by now, but this particular piece should be in in the next week).   Breaking changes after checkin will require high scrutiny.   Before we ship (next .NET Core is this June), we *could* entertain breaking changes but after that not really.   

This is part of the reason we are doing this in stages.  The Activity class is relatively simple and frankly does not do much (and does not have much 'policy' associated with it), so its relatively safe to check in.  Things like instrumenting HTTP, are next, but they are also not very 'breaking' so we have some time to fiddle with them as we experiment and test things end-to end.  

We appreciate your engagement.    It is certainly the case that we want a SHARED logging system, and we want interoperability with other systems.  Having your participation will help this and we want to thank you for your efforts here.

**Comment by vancem** [comment]
This has been implemented (pull request #15103)

**Comment by cwe1ss** [comment]
fyi, I posted some general questions about how this should be implemented in a tracer like Application Insights etc in https://github.com/dotnet/corefx/issues/16491
